### ISSUE TITLE
Modifier keys in menu items mistaken (Meta, Control)

### ISSUE THREAD

**Issue by Sciss** [mgarin/weblaf#117]
I just noticed that when running an application under OS X, the menu item short cuts show wrong modifiers. For example, if a short cut is <tt>Meta</tt> + <tt>A</tt>, this is only shown as <tt>A</tt>. In turn, if I use <tt>Ctrl</tt> + <tt>A</tt>, this is shown as <tt>⌘</tt> + <tt>A</tt> (using the command symbol rendering for Meta.)

**Comment by chaddaniels** [comment]
Is this fix still scheduled for 1.29?

**Comment by mgarin** [comment]
Yes, it is. All **bugs** which are set to v1.29 milestone will be fixed with v1.29 release. Most of features and enhancements which are set to v1.29 will also be included, but some of those might be re-scheduled for the next update depending on how this release goes.

**Comment by rquast** [comment]
Bump.. a customer informed me about this bug today. Is there a patch that can be released for it?

**Comment by rquast** [comment]
http://stackoverflow.com/questions/9780028/mac-keyboard-shortcuts-with-nimbus-laf - that solution worked for me as a workaround for each individual component that didn't work properly.

**Comment by mgarin** [comment]
I started to investigate this one and got kinda stuck. Unfortunately I'm not using mac a lot, so I'm not really sure how hotkeys should be handled there and which hotkeys for which application areas are better.

I have asked a few colleagues who are using Mac OS X a lot and they said that `ctrl` key is barely used or not used at all. I wonder whether it's true or depends on use cases.

Rght now there is an old workaround in WebLaF code that always handles `ctrl` in `HotkeyData` class as a menu shortcut key. That way it gets converted into `⌘` on Mac OS X, but stays `ctrl` on Windows and Linux. But that obviously disables usage of actual `ctrl` hotkey on Mac OS X at all.

So I'd really like to hear your comments about how different hotkeys should be acting on Mac OS X in different application areas. That would help me to correctly adjust the hotkeys handling within the library and avoid similar mistakes in future.

**Comment by Sciss** [comment]
The two issues:
- Binding of default action such as Cut/Copy/Paste should use platform specific conventions, so Meta-C/V/X on Mac, Ctrl-C/V/X on Linux and Windows (?)
- Rendering: You should always render Ctrl modifier as Ctrl character (^) and Meta modifier on OS X as ⌘. If I use Ctrl on Mac, it should still be drawn as Ctrl character. On the Mac you can use both Ctrl and Meta as modifiers, although primary menu actions (e.g. Meta-N/O/S for New/Open/Save) use meta.

**Comment by mgarin** [comment]
Sounds reasonable, but the problem is - Nimbus uses the same keys basic L&F provides. 
WebLaF uses those as well except a few components with slightly modified keys.

Here is an example of how L&F keystrokes are defined: 
```java
"TabbedPane.ancestorInputMap", new UIDefaults.LazyInputMap(new Object[] {
	"ctrl PAGE_DOWN", "navigatePageDown",
	"ctrl PAGE_UP", "navigatePageUp",
	"ctrl UP", "requestFocus",
	"ctrl KP_UP", "requestFocus",
}),
```
This specific part is the same for both Nimbus and WebLaF and I couldn't find how exactly hotkeys like `ctrl UP` are getting used as `meta UP` in the end. It doesn't seem that any particular code in Nimbus L&F takes care of that.

It seems that `MenuShortcut` class handles menu hotkeys properly (uses `ctrl` as `meta` for menu items), but there is nothing similar for components like `JTextField` - those seem to use hotkeys directly as they are provided, meaning they will use `ctrl` as `ctrl` and not `meta` if it was provided in key string.

Unfortunately I can't really run any tests on Mac OS X right now - will get my hands on a macbook only tomorrow, so right now I can only ask:
Do textfields with Nimbus L&F installed actually use `⌘+X/C/V` instead of `ctrl+X/C/V`?
By looking at the code it seems that textfields are using `ctrl+X/C/V` even on Mac OS X.

**Comment by Sciss** [comment]
I'm also not on Mac now, but from the code you posted that definitely will not used `⌘+X/C/V`, so perhaps my memory with Nimbus was wrong. In any case, since Mac has both Ctrl and Cmd, I don't think you need to make adjustments in every possible spot. For example those tabbed pane ones, I'm not even aware they exists. The crucial ones are the commonly used shortcuts that also appear in the edit menu: Meta-X/V/C/A/Z cut/copy/paste/select-all/undo. Redo has this stupid split across platforms - Meta-Shift-Z versus Ctrl-Y. Even on Linux I'm seeing applications randomly using one or the other. I think the nominal standard is Meta-Shift-Z for Mac _and Linux_, Ctrl-Y for Windows.

**Comment by mgarin** [comment]
> I'm also not on Mac now, but from the code you posted that definitely will not use ⌘+X/C/V, so perhaps my memory with Nimbus was wrong.

I guess I will check this out tomorrow then, maybe there is something I have missed in the code, some tricky workaround somewhere - Swing is full of those.

> In any case, since Mac has both Ctrl and Cmd, I don't think you need to make adjustments in every possible spot. For example those tabbed pane ones, I'm not even aware they existed. The crucial ones are the commonly used shortcuts that also appear in the edit menu: Meta-X/C/V/A/Z cut/copy/paste/select-all/undo.

The main question is whether I will have to do adjustments in L&F defaults or some tricky workarounds for specific components. I will need to test Nimbus first to see how it handles that stuff (if it does at all).

> Redo has this stupid split across platforms - Meta-Shift-Z versus Ctrl-Y. Even on Linux I'm seeing applications randomly using one or the other. I think the nominal standard is Meta-Shift-Z for Mac, Ctrl-Shift-Z for Linux, Ctrl-Y for Windows.

To be honest, `Ctrl+Y` is the worst hotkey ever. Most Windows applications I know are already using `Ctrl+Shift+Z` and sometimes you can find ones that work with both. Even Notepad is using `Ctrl+Shift+Z` already (and doesn't support `Ctrl+Y`).

I'd stick with `Ctrl(Meta)+Z` and `Ctrl(Meta)+Shift+Z` for all platforms for convenience reasons and just forget `Ctrl+Y` for good.
